modules/dnn/src/layers/reduce2_layer.cpp

```diff
@@ -0,0 +1,568 @@
+            case ReduceType::MEAN: strm << "MEAN"; break;
+            case ReduceType::SUM: strm << "SUM"; break;
+            case ReduceType::L1: strm << "L1"; break;
+            case ReduceType::L2: strm << "L2"; break;
+            case ReduceType::PROD: strm << "PROD"; break;
+            case ReduceType::SUM_SQUARE: strm << "SUM_SQUARE"; break;
+            case ReduceType::LOG_SUM: strm << "LOG_SUM"; break;
+            case ReduceType::LOG_SUM_EXP: strm << "LOG_SUM_EXP"; break;
+        }
+        strm << "\",\n";
+
+        prindent(strm, indent);
+        strm << "keepdims: " << (keepdims ? "true" : "false") << ",\n";
+
+        prindent(strm, indent);
+        strm << "noop_with_empty_axes: " << (noop_with_empty_axes ? "true" : "false") << ",\n";
+
+        prindent(strm, indent);
+        strm << "axes: [";
+        for (size_t i = 0; i < axes.size(); ++i) {
+            if (i > 0) strm << ", ";
+            strm << axes[i];
+        }
+        strm << "]\n";
+        return strm;
+    }
+
+    template <typename T, typename... Args>
+    inline void opDispatch(Args&&... args) {
+        using WT = typename WorkType<T>::type;
+        using FloatAcc = float;
+        using DoubleAcc = double;
+        using AccT = typename std::conditional<std::is_floating_point<WT>::value, FloatAcc, DoubleAcc>::type;
+        switch (reduce_type) {
+            case ReduceType::MAX:         ReduceInvoker<ReduceMax<T, WT, WT>>::run(std::forward<Args>(args)...);         break;
+            case ReduceType::MIN:         ReduceInvoker<ReduceMin<T, WT, WT>>::run(std::forward<Args>(args)...);         break;
+            case ReduceType::MEAN:        ReduceInvoker<ReduceMean<T, WT, AccT>>::run(std::forward<Args>(args)...);      break;
+            case ReduceType::SUM:         ReduceInvoker<ReduceSum<T, WT, AccT>>::run(std::forward<Args>(args)...);       break;
+            case ReduceType::L1:          ReduceInvoker<ReduceL1<T, WT, AccT>>::run(std::forward<Args>(args)...);        break;
+            case ReduceType::L2:          ReduceInvoker<ReduceL2<T, WT, AccT>>::run(std::forward<Args>(args)...);        break;
+            case ReduceType::PROD:        ReduceInvoker<ReduceProd<T, WT, WT>>::run(std::forward<Args>(args)...);        break;
+            case ReduceType::SUM_SQUARE:  ReduceInvoker<ReduceSumSquare<T, WT, AccT>>::run(std::forward<Args>(args)...); break;
+            case ReduceType::LOG_SUM:     ReduceInvoker<ReduceLogSum<T, WT, AccT>>::run(std::forward<Args>(args)...);    break;
+            case ReduceType::LOG_SUM_EXP: ReduceInvoker<ReduceLogSumExp<T, WT, AccT>>::run(std::forward<Args>(args)...); break;
+            default: CV_Error(Error::StsBadArg, "DNN/Reduce: Unsupported operation.");
+        }
+    }
+
+    template <typename... Args>
+    inline void typeDispatch(const int type, Args&&... args) {
+        switch (type) {
+            case CV_Bool: opDispatch<uint8_t>(std::forward<Args>(args)...); break;
```

Comment:
I'm not sure that CV_Bool means '0' or '1'. It can be '0' or '255' for example, or anything else. Probably we need to cast bool's to int's with `x != 0` code. That is, we need special handling for boolean type.